### PR TITLE
Backend: County-level Incarceration Data from Vera

### PR DESCRIPTION
**Pull request by benhammondmusic**
## Description

Uses csv file provided by Vera to produce data on `prison` and `jail` at the county level. Tables are kept separate by data type because they are from different years and therefor have different population comparison metrics. The Vera data is based on BJS reports but uses statistical analysis to normalize and infer county level data. The latest prison data in the open csv is from 2016, the latest jail data is from 2018. Incarceration data is largely out of date generally and this isn't really problematic compared to the recency of data in this field.

I don't believe we can generate any sort of combined incarceration metric, either disaggregated rates by demographic or the combined totals for children like we do for state/national reports from BJS. The data available on the [actual Vera site](https://trends.vera.org/) is newer and does include the combined incarceration rates; the data in the [public Vera github](https://github.com/vera-institute/incarceration-trends) csv does not. 

We may be able to include the raw count of children in jail at the county level (but not prison). Need further research in the methodology. 

## Motivation and Context

#1584 

## To Do

- [x] fix up tests to match pattern of BQ CALLS + SAMPLE BREAKDOWNS
- [x] add `Age` breakdown; decide if we can subdivide into `0-17` vs `18+` or just present totals
- [x] should we be `null`ing values below a certain threshold? we want to avoid presenting `0 per 100k` when it's not truly zero and simply a very small number that has been rounded down to `0` as an int

## How has this been tested?

- [x] overall test for `writeToBq()` function to confirm correct order of mocked calls
- [x] test coverage for all main breakdowns (eg `jail_age_county`, `prison_sex_county`), feeding in sample df and comparing results with an expected df
- [x] unit tests for any utility functions (might not have these are we are basically only using pandas fns to split, shape and recombine dfs

## Types of changes

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by joshzarrabi — APPROVED** [review]
Nice work. I do think we should standardize the county names based on the big query table but thats the only thing.

File: python/datasources/vera_incarceration_county.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+    Returns:
+        a dict mapping the data_type strings to the newly focused dfs for those types
+        { "prison": prison_df, "jail", jail_df}
+    """
+
+    df_jail = df.copy()
+    df_prison = df.copy()
+
+    # eliminate rows with unneeded years
+    df_jail = df_jail[df_jail["year"] == 2018].reset_index(drop=True)
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/vera_incarceration_county.py:181]: good idea

File: python/datasources/vera_incarceration_county.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+
+        # create and melt three partial dfs (to avoid column name collisions)
+        for property_type in [RAW, RATE, POP]:
+            partial_df = df.copy()
+            partial_df = generate_partial_breakdown(
+                partial_df, demo_type, data_type, property_type)
+            partial_breakdowns.append(partial_df)
+
+        # merge all the partial DFs for POP, RAW, RATE into a single DF per datatype/breakdown
+        breakdown_df = reduce(lambda x, y: pd.merge(
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/vera_incarceration_county.py:301]: Some of my finest stack-overflowing

File: python/datasources/vera_incarceration_county.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        breakdown_df = generate_pct_share_col_without_unknowns(
+            breakdown_df,
+            {POP: PCT_SHARE_COL_MAP[POP]},
+            demo_col,
+            all_val)
+
+        # add a column with the confined children
+        breakdown_df = pd.merge(breakdown_df, df_children, how="left", on=[
+            *GEO_COLS_TO_STANDARD.values(), demo_col])
+
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/vera_incarceration_county.py:326]: Ya I was thinking this yesterday with @eriwarr 's thing too

[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/vera_incarceration_county.py:326]: Proposed changes: #1639

File: python/datasources/vera_incarceration_county.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+    "white_pop_15to64": Race.WHITE_NH.value
+}
+
+SEX_POP_TO_STANDARD = {
+    "female_pop_15to64": Sex.FEMALE,
+    "male_pop_15to64": Sex.MALE
+}
+
+
+RACE_PRISON_RAW_COLS_TO_STANDARD = {
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/vera_incarceration_county.py:61]: It took some digging, but I traced back through and yes, they are using NH race groups.

Vera jail data (from Vera methodology, using BJS data):
Screen Shot 2022-06-27 at 10 22 27 PM

Vera prison data from National Corrections Reporting Program (NCRP):
prison NH

File: python/datasources/vera_incarceration_county.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+    def get_table_name():
+        return 'vera_incarceration_county'
+
+    def upload_to_gcs(self, _, **attrs):
+        raise NotImplementedError(
+            'upload_to_gcs should not be called for VeraIncarcerationCounty')
+
+    def write_to_bq(self, dataset, gcs_bucket, **attrs):
+
+        # TODO need to update naming scheme to be `vera_jail_data-age_county`
```
[the remark this answers is not in the document]
Comment (benhammondmusic) [reply python/datasources/vera_incarceration_county.py:246]: yup thanks

**Pull request merged**